lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -2976,7 +2977,54 @@ InferredAssociatedTypesByWitnesses
 ConformanceChecker::inferTypeWitnessesViaValueWitnesses(ValueDecl *req) {
   InferredAssociatedTypesByWitnesses result;
 
+  auto isExtensionUsableForInference = [&](ExtensionDecl *extension) -> bool {
+    // Assume concrete extensions we found witnesses in are always viable.
+    if (!extension->getExtendedType()->isAnyExistentialType())
```

Comment:
The assumption isn't quite true: constrained extensions of concrete types won't be able to provide suitable witnesses, so that `return true;` could more accurately be:

    return !extension->isConstrainedExtension();

for now. When conditional conformances come in, we'll need an "isAtLeastAsSpecializedAs" kind of check.